Transitive includes not handled correctly

Under certain setups imports of imported modules are not implicitly imported. It is important that this works 😅.

```.console
$ tail *.spicy
==> a.spicy <==
module a;

import spicy;

%byte-order=spicy::ByteOrder::Little;

public type A = unit {};

==> b.spicy <==
module b;

import a;
import spicy; # FIXME(bbannier): This should be implicit when importing `a`.

public type B = unit {
    a_: a::A;
};

==> c.spicy <==
module c;
import b;

==> d.spicy <==
module d;

import c;
```

When compiling both `c` and `d` this fails:

```.console
spicyc c.spicy d.spicy -j
/private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/d_dc75e53109471175-38c9e38328e44a4.cc:13:58: error: no member named 'a' in namespace '__hlt'
        std::optional<::hilti::rt::ValueReference<__hlt::a::A>> a_{};
                                                  ~~~~~~~^
```

As marked in the source for `b.spicy` there also seems to be an issue of `a` correctly exporting its dependency on `spicy`.

--- a/hilti/toolchain/include/compiler/unit.h
+++ b/hilti/toolchain/include/compiler/unit.h
@@ -7,6 +7,7 @@
 #include <functional>
 #include <memory>
 #include <string>
+#include <unordered_set>
 #include <utility>
 #include <vector>
 
@@ -172,8 +173,13 @@ public:
      */
     Result<CxxCode> cxxCode() const;
 
-    /** Returns the list of dependencies registered for the unit so far. */
-    const auto& dependencies() { return _dependencies; }
+    /**
+     * Returns the list of dependencies registered for the unit so far.
+     *
+     * @param recursive if true, return the transitive closure of all
+     * dependent units, vs just direct dependencies of the current unit
+     */
+    std::vector<std::weak_ptr<Unit>> dependencies(bool recursive = false) const;
 
     /** Removes any dependencies registered for the unit so far. */
     void clearDependencies() { _dependencies.clear(); };
@@ -384,6 +390,9 @@ private:
 
     // Recursively destroys the module's AST.
     void _destroyModule();
+
+    // Helper for depencencies() to recurse.
+    void _recursiveDependencies(std::vector<std::weak_ptr<Unit>>* dst, std::unordered_set<const Unit*>* seen) const;
 
     // Parses a source file with the appropriate plugin.
     static Result<hilti::Module> _parse(const std::shared_ptr<Context>& context,

--- a/hilti/toolchain/src/compiler/unit.cc
+++ b/hilti/toolchain/src/compiler/unit.cc
@@ -266,7 +266,7 @@ Result<Nothing> Unit::codegen() {
     //
     // TODO(robin): Would be nice if we had a "cheap" compilation mode that
     // only generated declarations.
-    for ( const auto& unit : dependencies() ) {
+    for ( const auto& unit : dependencies(true) ) {
         HILTI_DEBUG(logging::debug::Compiler, fmt("importing declarations from module %s", unit.lock()->id()));
         auto other = detail::CodeGen(context()).compileModule(unit.lock()->module(), unit.lock().get(), false);
         c->importDeclarations(*other);
@@ -305,6 +305,32 @@ Result<CxxCode> Unit::cxxCode() const {
         return result::Error("errors during prototype creation");
 
     return CxxCode{_cxx_unit->moduleID(), cxx};
+}
+
+void Unit::_recursiveDependencies(std::vector<std::weak_ptr<Unit>>* dst, std::unordered_set<const Unit*>* seen) const {
+    // This uses two vectors because the weak_ptr are a bit tough to work with,
+    // in particular they can't be compared through std::find().
+
+    for ( const auto& d : _dependencies ) {
+        auto dptr = d.lock().get();
+
+        if ( seen->find(dptr) != seen->end() )
+            continue;
+
+        dst->push_back(d);
+        seen->insert(dptr);
+        dptr->_recursiveDependencies(dst, seen);
+    }
+}
+
+std::vector<std::weak_ptr<Unit>> Unit::dependencies(bool recursive) const {
+    if ( ! recursive )
+        return _dependencies;
+
+    std::vector<std::weak_ptr<Unit>> deps;
+    std::unordered_set<const Unit*> seen;
+    _recursiveDependencies(&deps, &seen);
+    return deps;
 }
 
 bool Unit::addDependency(std::shared_ptr<Unit> unit) {

--- a/tests/Baseline/hilti.ast.imported-id/output
+++ b/tests/Baseline/hilti.ast.imported-id/output
@@ -144,6 +144,7 @@
 [debug/compiler]     dependencies: Foo
 [debug/compiler]   compiling module Foo to C++
 [debug/compiler]     importing declarations from module Bar
+[debug/compiler]     importing declarations from module Foo
 [debug/compiler]     finalizing module Foo
 // Begin of Foo (from "foo.hlt")
 // Compiled by HILTI version X.X.X

--- a/tests/spicy/types/unit/transitive-imports.spicy
+++ b/tests/spicy/types/unit/transitive-imports.spicy
@@ -1,0 +1,34 @@
+# @TEST-EXEC: spicyc -d -j c.spicy d.spicy
+#
+# @TEST-DOC: Regression test for #1057 to check C++ type emission for transitive imports
+
+@TEST-START-FILE a.spicy
+module a;
+
+import spicy;
+
+%byte-order=spicy::ByteOrder::Little;
+
+public type A = unit {};
+@TEST-END-FILE
+
+@TEST-START-FILE b.spicy
+module b;
+
+import a;
+
+public type B = unit {
+    a_: a::A;
+};
+@TEST-END-FILE
+
+@TEST-START-FILE c.spicy
+module c;
+import b;
+@TEST-END-FILE
+
+@TEST-START-FILE d.spicy
+module d;
+
+import c;
+@TEST-END-FILE

